include/xtensor-sparse/xcsr_scheme.hpp

```diff
@@ -0,0 +1,463 @@
+        using pointer = typename storage_type::pointer;
+
+        using iterator = xcsr_scheme_iterator<self_type>;
+        using const_iterator = xcsr_scheme_iterator<const self_type>;
+
+        xcsr_scheme(std::size_t size);
+
+        const position_type& position() const;
+        const coordinate_type& coordinate() const;
+        const storage_type& storage() const;
+
+        storage_type& storage();
+
+        pointer find_element(const index_type& index);
+        void insert_element(const index_type& index, const_reference value);
+        void remove_element(const index_type& index);
+
+        template <class strides_type, class shape_type>
+        void update_entries(const strides_type& old_strides,
+                            const strides_type& new_strides,
+                            const shape_type& new_shape);
+
+        iterator begin();
+        iterator end();
+        const_iterator begin() const;
+        const_iterator end() const;
+        const_iterator cbegin() const;
+        const_iterator cend() const;
+
+    private:
+
+        position_type m_pos;
+        coordinate_type m_coords;
+        storage_type m_storage;
+
+        friend class xcsr_scheme_iterator<self_type>;
+        friend class xcsr_scheme_iterator<const self_type>;
+    };
+
+    /************************************
+     * xcsr_scheme_iterator declaration *
+     ************************************/
+
+    namespace detail
+    {
+        template <class scheme>
+        struct xcsr_scheme_storage_type
+        {
+            using storage_type = typename scheme::storage_type;
+            using value_iterator = typename storage_type::iterator;
+        };
+
+        template <class scheme>
+        struct xcsr_scheme_storage_type<const scheme>
+        {
+            using storage_type = typename scheme::storage_type;
+            using value_iterator = typename storage_type::const_iterator;
+        };
+
+        template <class scheme>
+        struct xcsr_scheme_iterator_types : xcsr_scheme_storage_type<scheme>
+        {
+            using base_type = xcsr_scheme_storage_type<scheme>;
+            using index_type = typename scheme::index_type;
+            using position_type = typename scheme::position_type;
+            using position_iterator = typename position_type::const_iterator;
+            using coordinate_type = typename scheme::coordinate_type;
+            using coordinate_iterator = typename coordinate_type::const_iterator;
+            using value_iterator = typename base_type::value_iterator;
+            using value_type = typename value_iterator::value_type;
+            using reference = typename value_iterator::reference;
+            using pointer = typename value_iterator::pointer;
+            using difference_type = typename value_iterator::difference_type;
+        };
+    }
+
+    template <class scheme>
+    class xcsr_scheme_iterator: public xtl::xrandom_access_iterator_base<xcsr_scheme_iterator<scheme>,
```

Comment:
You can simplify by inheriting from `xtl::xrandom_acces_iterator_base3<xcsr_scheme_iterator<scheme>, xcsr_scheme_iterator_types<scheme>>`